src_c/simd_collisions_avx2.c

```diff
@@ -0,0 +1,172 @@
+    }
+
+    double t = DBL_MAX;
+
+    // here we know that there is at least one intersection so
+    // we search for the smallest t value that still meets the above conditions
+    int i = 0;
+    for (i = 0; i < 4; i++) {
+        if (((double *)&t_u_256d)[i]) {
+            t = MIN(t, ((double *)&t_256d)[i]);
+        }
+    }
+
+    // outputs
+    if (T)
+        *T = t;
+    if (X)
+        *X = line->x1 + t * (line->x2 - line->x1);
+    if (Y)
+        *Y = line->y1 + t * (line->y2 - line->y1);
+
+    return 1;
+}
+#endif /* ~AVX2_IS_SUPPORTED */
+
+#if AVX2_IS_SUPPORTED
+PG_FORCEINLINE static int
+pgCollision_RectLine_avx2(SDL_Rect *rect, pgLineBase *line)
+{
+    double Rx = (double)rect->x;
+    double Ry = (double)rect->y;
+    double Rw = (double)rect->w;
+    double Rh = (double)rect->h;
+
+    // here we start to setup the variables
+    __m256d x1_256d = _mm256_set1_pd(line->x1);
+    __m256d y1_256d = _mm256_set1_pd(line->y1);
+    __m256d x2_256d = _mm256_set1_pd(line->x2);
+    __m256d y2_256d = _mm256_set1_pd(line->y2);
+    __m256d x3_256d = _mm256_set_pd(Rx, Rx, Rx, Rx + Rw);
+    __m256d y3_256d = _mm256_set_pd(Ry, Ry, Ry + Rh, Ry);
+    __m256d x4_256d = _mm256_set_pd(Rx + Rw, Rx, Rx + Rw, Rx + Rw);
+    __m256d y4_256d = _mm256_set_pd(Ry, Ry + Rh, Ry + Rh, Ry + Rh);
+
+    // here we calculate the differences between the the coords of the points
+    __m256d x1_m_x2_256d = _mm256_sub_pd(x1_256d, x2_256d);
+    __m256d y3_m_y4_256d = _mm256_sub_pd(y3_256d, y4_256d);
+    __m256d y1_m_y2_256d = _mm256_sub_pd(y1_256d, y2_256d);
+    __m256d x3_m_x4_256d = _mm256_sub_pd(x3_256d, x4_256d);
+
+    // we calculate the denominator of the equations
+    __m256d den_256d =
+        _mm256_sub_pd(_mm256_mul_pd(x1_m_x2_256d, y3_m_y4_256d),
+                      _mm256_mul_pd(y1_m_y2_256d, x3_m_x4_256d));
+
+    // if the denominator is 0 then the line is parallel to the other line
+    // in this occasion this can't be true here as a line will never be
+    // parallel to all four sides of a rectangle
+    __m256d den_zero_256d =
+        _mm256_cmp_pd(den_256d, _mm256_setzero_pd(), _CMP_EQ_OQ);
+
+    // we dont want to cause any floating point errors by dividing by 0
+    // so we set the ones that are equal to 0 to 1
+    den_256d = _mm256_or_pd(den_zero_256d, den_256d);
+
+    // we calculate the rest of the differences between the coords of the
+    // points
+    __m256d x1_m_x3_256d = _mm256_sub_pd(x1_256d, x3_256d);
+    __m256d y1_m_y3_256d = _mm256_sub_pd(y1_256d, y3_256d);
+
+    // calculate the t values
+    __m256d t_256d = _mm256_sub_pd(_mm256_mul_pd(x1_m_x3_256d, y3_m_y4_256d),
+                                   _mm256_mul_pd(y1_m_y3_256d, x3_m_x4_256d));
+    t_256d = _mm256_div_pd(t_256d, den_256d);
+
+    // calculate the u values
+    __m256d u_256d = _mm256_sub_pd(_mm256_mul_pd(x1_m_x2_256d, y1_m_y3_256d),
+                                   _mm256_mul_pd(y1_m_y2_256d, x1_m_x3_256d));
+    u_256d =
```

Comment:
I have not fully thought this through, but
rather than multiplying this by -1 to negate it, could you instead flip the comparison operators on what you compare this too?

Reply:
the way to do this is to compare for smaller or equal to 0 and larger or equal to -1, i dont think this gives us any speed improvement as instead of calling one AVX2 SIMD intrinsic we create an extra register with -1.
PS. this idea of just flipping the comparisons wont work cause we will go from `0 <= u <= 1` to `1 <= u <= 0` thus rendering it invalid.